app/controllers/session_controller.rb

```diff
@@ -233,6 +238,14 @@ def create
       return admin_not_allowed_from_ip_address(user)
     end
 
+    if user.totp_enabled?
+      unless user.authenticate_otp(params[:second_factor_token])
+        secure_session["second-factor-#{request.remote_ip}"] = "true"
```

Comment:
this feels risky cause ip can change between requests especially in corporates behind a big NAT

Reply:
The only part where this is used is to identify which rate limiter to use -- the "login" one or the "second factor token" one.

An alternative to using this method is to detect whether the second_factor_token parameter is in the request. It could trivially give attackers a few more attempts against a different rate limiter, but it would still not allow a crazy amout of brute forcing.